python/pyspark/sql/types.py

```diff
@@ -194,16 +194,7 @@ def fromDDL(cls, ddl: str) -> "DataType":
         >>> DataType.fromDDL("b: string, a: int")
         StructType([StructField('b', StringType(), True), StructField('a', IntegerType(), True)])
         """
-        from pyspark.sql import SparkSession
-        from pyspark.sql.functions import udf
-
-        # Intentionally uses SparkSession so one implementation can be shared with/without
-        # Spark Connect.
-        schema = (
-            SparkSession.active().range(0).select(udf(lambda x: x, returnType=ddl)("id")).schema
-        )
-        assert len(schema) == 1
-        return schema[0].dataType
+        return _parse_datatype_string(ddl)
```

Comment:
Why is this change needed?

Reply:
If we want to do something like `fromDDL("v variant")`, `fromDDL` actually calls a `udf` just to get the output schema. However, this PR disables variant output UDFs during planning, so we have to get the data type another way

I've discussed a bit with @HyukjinKwon offline last night - he suggested that we instead block variant ser/de, however I took a second look today and, unless I'm missing something, I believe this route is preferable:

- This will fail earlier during planning rather than execution
- Blocking during ser/de may require more code changes, because pandas udfs don't seem to have the same codepath as non-pandas udfs.
- the `_parse_datatype_string` also was made to work with spark connect (which was @HyukjinKwon's initial concern) in [this](https://github.com/apache/spark/commit/fa8aa571ad18441622bb7e3ac66032ab9e7cbc0a) pr

Reply:
Can we make sure the behaivor of `_parse_datatype_string` is the same as the original `fromDDL`? My concern is that this might introduce unintentional behavior change for a public API. 
What's the error message if we do  `fromDDL(a variant)` without this change?

Reply:
yeah makes sense. The function comment mentions that this was added for Spark 4.0.0 (which won't be GA until 2025 IIUC?) so I thought the risk of a breaking change for a public API would be a bit lower.

The error message without changing `fromDDL` for something like "a int v variant" would be
```
org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.UNSUPPORTED_UDF_OUTPUT_TYPE] Cannot resolve "<lambda>(id)" due to data type mismatch: UDFs do not support 'STRUCT<a: INT, v: VARIANT>' as an output data type. SQLSTATE: 42K09
```
which is pretty confusing for `fromDDL` in my opinion (i.e. why does `fromDDL` need to call a UDF in the first place?)

It seems that the UDF `returnType` implementation directly calls `_parse_datatype_string` [here](https://github.com/apache/spark/blob/master/python/pyspark/sql/udf.py#L370).

I added an additional test to test that the old udf behavior matches the new behavior